src/manifest.rs

```diff
@@ -267,11 +269,23 @@ impl Manifest {
             // to
             // alias = { version = "0.2", package = "a" }
             if let Some(renamed) = dep.rename() {
-                let old_copy = table[&dep.name].clone();
-                table[renamed] = old_copy;
-                table[&dep.name] = toml_edit::Item::None;
+                table[renamed] = dep_item.clone();
+                table[&dep_name] = toml_edit::Item::None;
+                dep_name = renamed.to_owned();
+            } else if dep.name != dep_name {
+                // if `dep` had been renamed in the manifest,
+                // and is not rename in the `add` command,
+                // we need to remove the old entry and insert a new one
+                // e.g. from
+                // alias = { version = "0.1", package = "a" }
+                // to
+                // a = "0.2"
+                table[&dep_name] = toml_edit::Item::None;
```

Comment:
Hmm, now that I think about it, it can be more tricky if we want to preserve other fields in the table like e.g. `features` or `optional`. Ideally, we only want to remove it if it's length is 2, but that would complicate the code, argh. @killercup wdyt?

Reply:
note that removing an item also deletes the comments associated with it (on the right and above)

Reply:
It sounds like this would be an `upgrade` operation if we merge in that context. Maybe for `add`, it would make more sense to stop with an error if length > 2, otherwise we can safely drop & re-create ?

Reply:
we already support that e.g. with `optional` field, try adding an optional dependency and then `cargo add` the same dependency

Reply:
Our current behavior for adding a table (if there are no renames) is to merge tables the old and the new table into one. It would be surprising if we change that if renames are involved.